contracts/early_bird_special/lib.rs

```diff
@@ -0,0 +1,286 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+/// This is the EarlyBirdSpecial
+///
+/// Larger rewards are distributed for engaging in the game as early on as possible:
+/// user_score = deadline - now
+/// On the other hand ThePressiah (the last player to click) gets 50% of the token pool, which creates two competing strategies.
+
+#[ink::contract(env = button::ButtonGameEnvironment)]
+mod early_bird_special {
+
+    use access_control::{traits::AccessControlled, Role, ACCESS_CONTROL_PUBKEY};
+    use button::{
+        ButtonData, ButtonGame, ButtonGameEnvironment, Error, IButtonGame, Result, Score,
+    };
+    use button_token::{BALANCE_OF_SELECTOR, TRANSFER_SELECTOR};
+    use ink_env::Error as InkEnvError;
+    use ink_lang::{
+        codegen::{initialize_contract, EmitEvent},
+        reflect::ContractEventBase,
+    };
+    use ink_prelude::{format, vec::Vec};
+    use ink_storage::traits::SpreadAllocate;
+
+    /// Event type
+    type Event = <EarlyBirdSpecial as ContractEventBase>::Type;
+
+    /// Event emitted when TheButton is created
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct ButtonCreated {
+        #[ink(topic)]
+        button_token: AccountId,
+        start: BlockNumber,
+        deadline: BlockNumber,
+    }
+
+    /// Event emitted when account is whitelisted to play the game
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct AccountWhitelisted {
+        #[ink(topic)]
+        player: AccountId,
+    }
+
+    /// Event emitted when account is blacklisted from playing the game
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct AccountBlacklisted {
+        #[ink(topic)]
+        player: AccountId,
+    }
+
+    /// Event emitted when TheButton is pressed
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct ButtonPressed {
+        #[ink(topic)]
+        by: AccountId,
+        when: BlockNumber,
+    }
+
+    /// Event emitted when button's death is triggered
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct ButtonDeath {}
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct EarlyBirdSpecial {
+        data: ButtonData,
+    }
+
+    impl AccessControlled for EarlyBirdSpecial {
+        type ContractError = Error;
+    }
+
+    impl ButtonGame for EarlyBirdSpecial {
+        fn get(&self) -> &ButtonData {
+            &self.data
+        }
+
+        fn get_mut(&mut self) -> &mut ButtonData {
+            &mut self.data
+        }
+
+        fn score(&self, now: BlockNumber) -> Score {
+            let deadline = ButtonGame::deadline(self);
+            deadline - now
+        }
+    }
+
+    // becasue ink! does not allow generics or trait default implementations
+    impl IButtonGame for EarlyBirdSpecial {
+        #[ink(message)]
+        fn is_dead(&self) -> bool {
+            let now = Self::env().block_number();
+            ButtonGame::is_dead(self, now)
+        }
+
+        #[ink(message)]
+        fn press(&mut self) -> Result<()> {
```

Comment:
I think that there is an opportunity to improve UX here by adding a parameter `press_number: u32` here that would be passed within the transaction. 
The meaning of this `press_number` argument within the `press` transaction is the following "I'm willing to press the button with the intention of my press being the number `press_number` counting from the start of the game." The contract should maintain in its state a variable `number_pressed` (or some such), and whenever a transaction comes saying press with some outdated number, then just ignore it (I mean, the press does not happen).

This helps a lot when lots of users are trying to land a button press at a particular moment in time (especially when competing with bots). If 5 of them send a tx at the same time with the current solution, then 1 of them will succeed and 4 of them will "waste" their press (and will be angry). While in what I propose, the one guy will win, but the remaining guys presses will be canceled, so they will keep their chance to score high.

Reply:
This is another great comment, that made me re-think how the game should work.

As it was, if multiple account recorded their press at the same time, they would all get recorded (so not wasted per se) with the same score. When it comes to the Pressiah than effectively the account that submitted the last tx processed for the last block of the Buttons life becomes the Pressiah - which could be seen as counter-intuitive.

What I propose (subject to discussion OC) is this:
- the block number (the `now`) becomes a type of a nonce for the `press` transactions. 
- in each block there can be only one such tx processed, so there is also a per-block competition to be that one included in it, in addition to the overall game competition
- if you were late, the tx returns an error, but you get to try again (your press is not recorded this time)

Reply:
I think the version you propose is good. Although I would still argue that for the `back_to_the_future` game this can lead to some unpleasant surprises for the users. For instance
-- The button is close to dead -- 5 seconds from the end
-- Two users click -- for one of them the transaction lands in block 896, and for the other 897
-- The second user is pissed, because it wanted ~900 points and not 1 xD

Reply:
Also a general consideration -- relying on block numbers instead of timestamps (which ink exposes I would presume?) might sometimes have bad effects. If we end up relying on block numbers, we need to make sure the UI displays this accordingly. We should ideally not assume 1 block=1 sec then.

Reply:
ad 1) Unless I'm missing or did not understand something, if we allow for passing a "desired" number in a press tx argument, then the game becomes very "gameable".
Taking for example the `back_to_the_future` game, where the reward is bigger the closer to the end of the Buttons life it is (even ignoring the whoel pressiah business).
- lets say the game starts in block number 0 and lives for 900 blocks
- just in the block 0 you submit press(900) and it gets into the block - your score becomes 900-0 and you already "won"

Regardless of the logic of the game it is now always desired to be the first one to play, because you get to submit the winning transaction right of the bat. I mean others can still play for the scrappy seconds, but for example the lucrative Pressiah status is taken right away :)

ad 2) timestamps are not exposed, just block number - maybe instead of a timer the UI should display block numbers counting down to the game end?

Reply:
1) That's not what I meant :) 
What I imagine is that you should initialize the state of the game with `number_pressed=0`
Then whoever sends `press(0)` first, presses the button for the first time and `number_pressed` gets incremented, then we expect `press(1)`, etc.
More generally: in the state `number_pressed=k` all `press(x)` transactions with `x!=k` are ignored, and only the first one with `x=k` gets accepted.

It's just "adding a nonce" being a counter of presses, so that people don't get screwed when pressing at roughly the same time.
2) OK, in that case I would be in favor of the UI measuring time in blocks and not in seconds...